bookkeeper-server/src/main/java/org/apache/bookkeeper/client/UpdateLedgerOp.java

```diff
@@ -81,174 +74,100 @@ public UpdateLedgerOp(final BookKeeper bkc, final BookKeeperAdmin admin) {
         final AtomicInteger issuedLedgerCnt = new AtomicInteger();
         final AtomicInteger updatedLedgerCnt = new AtomicInteger();
-        final Future<?> updateBookieCb = executor.submit(new Runnable() {
-
-            @Override
-            public void run() {
-                updateLedgers(oldBookieId, newBookieId, rate, limit, progressable);
-            }
-
-            private void updateLedgers(final BookieSocketAddress oldBookieId, final BookieSocketAddress newBookieId,
-                    final int rate, final int limit, final UpdateLedgerNotifier progressable) {
-                try {
-                    final AtomicBoolean stop = new AtomicBoolean(false);
-                    final Set<Long> outstandings = Collections.newSetFromMap(new ConcurrentHashMap<Long, Boolean>());
-                    final RateLimiter throttler = RateLimiter.create(rate);
-                    final Iterator<Long> ledgerItr = admin.listLedgers().iterator();
-                    final CountDownLatch syncObj = new CountDownLatch(1);
-
-                    // iterate through all the ledgers
-                    while (ledgerItr.hasNext() && !stop.get()) {
-                        // throttler to control updates per second
-                        throttler.acquire();
-
-                        final Long lId = ledgerItr.next();
-                        final ReadLedgerMetadataCb readCb = new ReadLedgerMetadataCb(bkc, lId, oldBookieId,
-                                newBookieId);
-                        outstandings.add(lId);
-
-                        FutureCallback<Void> updateLedgerCb = new UpdateLedgerCb(lId, stop, issuedLedgerCnt,
-                                updatedLedgerCnt, outstandings, syncObj, progressable);
-                        Futures.addCallback(readCb.getFutureListener(), updateLedgerCb);
-
-                        issuedLedgerCnt.incrementAndGet();
-                        if (limit != Integer.MIN_VALUE && issuedLedgerCnt.get() >= limit || !ledgerItr.hasNext()) {
-                            stop.set(true);
+        final CompletableFuture<Void> finalPromise = new CompletableFuture<>();
+        final Set<CompletableFuture<?>> outstanding =
+            Collections.newSetFromMap(new ConcurrentHashMap<CompletableFuture<?>, Boolean>());
+        final RateLimiter throttler = RateLimiter.create(rate);
+        final Iterator<Long> ledgerItr = admin.listLedgers().iterator();
+
+        // iterate through all the ledgers
+        while (ledgerItr.hasNext() && !finalPromise.isDone()
+               && (limit == Integer.MIN_VALUE || issuedLedgerCnt.get() < limit)) {
+            // throttler to control updates per second
+            throttler.acquire();
+
+            final long ledgerId = ledgerItr.next();
+            issuedLedgerCnt.incrementAndGet();
+
+            GenericCallbackFuture<LedgerMetadata> readPromise = new GenericCallbackFuture<>();
+            lm.readLedgerMetadata(ledgerId, readPromise);
```

Comment:
I think it will be much clearer (and easier to maintain in future) if you call lm.readLedgerMetadata at the end of the sequence (after adding the promise to outstanding list).

```
GenericCallbackFuture<LedgerMetadata> readPromise = ...;
CompletableFuture<LedgerMetadata> writePromise = readPromise.thenCompose(...);
// register a completion callback when updating ledger metadata completes
writePromise.whenComplete(...);
// add outstanding write to outstanding list
outstanding.add(writePromise);
// trigger the read
lm.readLedgerMetadata(ledgerId, readPromise);
```

Reply:
I disagree, and this is of the things that annoys me about callbacks. 

When reading code, it makes most sense that the action occur in the order they occur on the page. By putting readLedgerMetadata at the end, you are breaking that order. Consider how this would look if instead of taking a promise, readLedgerMetadata would return a future (something I've been thinking about doing).

Reply:
sure